test/integration/agent/route_test.go

```diff
@@ -380,13 +380,66 @@ func TestAddAndDeleteRoutes(t *testing.T) {
 		output, err := ExecOutputTrim(fmt.Sprintf("ip route show table 0 exact %s", peerCIDR))
 		assert.NoError(t, err)
 		assert.Equal(t, "", output, "expected no routes to %s", peerCIDR)
-
 		entries, err = ipset.ListEntries("ANTREA-POD-IP")
 		assert.NoError(t, err, "list ipset entries failed")
 		assert.NotContains(t, entries, tc.peerCIDR, "entry should not be in ipset")
 	}
 }
 
+func TestSyncRoutes(t *testing.T) {
+	skipIfNotInContainer(t)
+	gwLink := createDummyGW(t)
+	defer netlink.LinkDel(gwLink)
+
+	tcs := []struct {
+		// variations
+		mode     config.TrafficEncapModeType
+		nodeName string
+		peerCIDR string
+		peerIP   net.IP
+		// expectations
+		uplink netlink.Link // indicates outbound of the route.
+	}{
+		{mode: config.TrafficEncapModeEncap, nodeName: "node0", peerCIDR: "10.10.20.0/24", peerIP: localPeerIP, uplink: gwLink},
+		{mode: config.TrafficEncapModeNoEncap, nodeName: "node1", peerCIDR: "10.10.30.0/24", peerIP: localPeerIP, uplink: nodeLink},
+		{mode: config.TrafficEncapModeNoEncap, nodeName: "node2", peerCIDR: "10.10.40.0/24", peerIP: remotePeerIP, uplink: nil},
+		{mode: config.TrafficEncapModeHybrid, nodeName: "node3", peerCIDR: "10.10.50.0/24", peerIP: localPeerIP, uplink: nodeLink},
+		{mode: config.TrafficEncapModeHybrid, nodeName: "node4", peerCIDR: "10.10.60.0/24", peerIP: remotePeerIP, uplink: gwLink},
+	}
+
+	for _, tc := range tcs {
+		t.Logf("Running test with mode %s peer cidr %s peer ip %s node config %s", tc.mode, tc.peerCIDR, tc.peerIP, nodeConfig)
+		routeClient, err := route.NewClient(serviceCIDR, &config.NetworkConfig{TrafficEncapMode: tc.mode}, false)
+		assert.NoError(t, err)
+		err = routeClient.Initialize(nodeConfig, func() {})
+		assert.NoError(t, err)
+
+		_, peerCIDR, _ := net.ParseCIDR(tc.peerCIDR)
+		nhCIDRIP := ip.NextIP(peerCIDR.IP)
+		assert.NoError(t, routeClient.AddRoutes(peerCIDR, tc.nodeName, tc.peerIP, nhCIDRIP), "adding routes failed")
+
+		if !tc.mode.NeedsEncapToPeer(tc.peerIP, nodeConfig.NodeIPAddr) && tc.mode.NeedsRoutingToPeer(tc.peerIP, nodeConfig.NodeIPAddr) {
```

Comment:
Isn't the first condition redundant? And it doesn't seem necessary to add a case that will be skipped. But maybe it could check whatever the original route is (a valid route or nil), after removing the route and a sync loop, it can get the same route, instead of checking the route is not empty

Reply:
The first condition is redundant. But the second is to avoid error when executing "ip route del" on nonexistent route since in that case the route will not be added according to https://github.com/vmware-tanzu/antrea/blob/3335e734071894f70a9ab33c734799f56c049689/pkg/agent/route/route_linux.go#L539

Reply:
Then why the case is even added to TestSyncRoute if it's skipped anyway?  I think it should either remove the 3rd case and the mode check, or be more generic: only call "ip route del" if "expOutput" is not empty, so that the test works for all cases and don't have to check mode.

Reply:
```		expOutput, err := exec.Command("bash", "-c", listCmd).Output()
		assert.NoError(t, err, "error executing ip route command: %s", listCmd)

                if len(expOutput) > 0 {
		        delCmd := fmt.Sprintf("ip route del %s", expOutput)
		        _, err = exec.Command("bash", "-c", delCmd).Output()
		        assert.NoError(t, err, "error executing ip route command: %s", delCmd)
                }